### PR TITLE
wr-203:  Add achievements checking functionality

### PR DESCRIPTION
**Pull request by hnoievets**
**I implemented this task in another way.** 
In this branch I don't use `achievements` from the slice, and `ownAchievements` looks like this:
image

Because I believe, filtering user achievements during calculating progress, and then fetching those user achievements separately from a client side is unnecessary work, and I don't understand where to store `progress` on FE if we would have `state.achievements` - base achievements and `state.ownAchievements` - objects with: id, userId, and achievementId 😐

- [x] Create users_achievements (with schema provided from readme) table (user_id | achievement_id cols pair should be unique)
- [x]  Add 2 columns for achievements table (breakpoint with count of items to achieve it and reference_table with table name for referencing achievement)
- [x]  On saving Article call achievements service to check wether any achievement breakpoint was accomplished:
- If breakpoint is accomplished - then make record in users_achievements table
- [x]  On saving Comment - same way, but for comments table
- If breakpoint is accomplished - then make record in users_achievements table
- [x]  On user profile - fetch user achievements and show the actual ones
Also implemented calculating achievement progress.

image

image

### CODE REVIEW

File: backend/src/db/migrations/20230913114720_add_breakpoint_and_reference_table_columns_to_achievements.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+  ARTICLES: 'articles',
+  COMMENTS: 'comments',
+} as const;
+
+const ColumnName = {
+  BREAKPOINT: 'breakpoint',
+  REFERENCE_TABLE: 'reference_table',
+} as const;
+
+const DataToUpdate = [
```
Comment (andanf-e) [inline backend/src/db/migrations/20230913114720_add_breakpoint_and_reference_table_columns_to_achievements.ts]: As it's a constant
```suggestion
const DATA_TO_UPDATE = [
```

File: backend/src/db/migrations/20230913114720_add_breakpoint_and_reference_table_columns_to_achievements.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
+import { type Knex } from 'knex';
+
+const TableName = {
+  ACHEVEMENTS: 'achievements',
```
Comment (andanf-e) [inline backend/src/db/migrations/20230913114720_add_breakpoint_and_reference_table_columns_to_achievements.ts]: typo
```suggestion
  ACHIEVEMENTS: 'achievements',
```

File: backend/src/packages/achievements/achievement.entity.ts [hunk 3]
```diff
@@ -14,42 +14,58 @@ class AchievementEntity implements IEntity {
 
   private 'description': string;
 
+  private 'breakpoint': number;
+
+  private 'referenceTable': string;
```
Comment (andanf-e) [inline backend/src/packages/achievements/achievement.entity.ts]:
```suggestion
  private 'referenceTable': typeof DatabaseTableName.ARTICLES | typeof DatabaseTableName.COMMENTS;
```

File: backend/src/packages/achievements/achievement.model.ts [hunk 4]
```diff
@@ -7,6 +7,8 @@ class AchievementModel extends AbstractModel {
   public 'key': string;
   public 'name': string;
   public 'description': string;
+  public 'breakpoint': number;
+  public 'referenceTable': string;
```
Comment (andanf-e) [inline backend/src/packages/achievements/achievement.model.ts]:
```suggestion
  public 'referenceTable': typeof DatabaseTableName.ARTICLES | typeof DatabaseTableName.COMMENTS;
```

File: backend/src/packages/achievements/achievement.repository.ts [hunk 5]
```diff
@@ -29,14 +40,88 @@ class AchievementRepository implements IRepository {
...
+
+    return achievements.map((userAchievement) => ({
+      id: userAchievement.id,
+      userId: userAchievement.userId,
+      achievementId: userAchievement.achievementId,
+    }));
+  }
+
+  public async findBy(
+    filters: Partial<Achievement> = {},
```
Comment (andanf-e) [inline backend/src/packages/achievements/achievement.repository.ts]: As they are only ones used
```suggestion
    filters: Pick<Achievement, 'breakpoint' | 'referenceTable'>,
```

File: backend/src/packages/achievements/achievement.repository.ts [hunk 6]
```diff
@@ -1,13 +1,24 @@
 import { type IRepository } from '~/libs/interfaces/repository.interface.js';
+import { type DatabaseTableName } from '~/libs/packages/database/database.js';
+import { type ValueOf } from '~/libs/types/types.js';
 
 import { AchievementEntity } from './achievement.entity.js';
 import { type AchievementModel } from './achievement.model.js';
+import { type Achievement, type UserAchievement } from './libs/types/types.js';
+import { type UserAchievementModel } from './user-achievements.model.js';
+
+type referenceTableType = ValueOf<typeof DatabaseTableName>;
```
Comment (andanf-e) [inline backend/src/packages/achievements/achievement.repository.ts]: 1. Types should be named with capitalized letter
2. If you're using in through app - then add it to common types
3. I believe we should restrict it with currently maintained reference tables (`DatabaseTableName.ARTICLES` and `DatabaseTableName.COMMENTS`)

File: backend/src/packages/achievements/achievement.repository.ts [hunk 7]
```diff
@@ -29,14 +40,88 @@ class AchievementRepository implements IRepository {
...
     return Promise.resolve(null);
   }
 
+  public async getCountsInReferenceTables(
+    userId: number,
+  ): Promise<{ [key: string]: number }[]> {
+    const relatedTableNames = await this.achievementModel
+      .query()
+      .select('referenceTable')
+      .groupBy('referenceTable')
```
Comment (andanf-e) [inline backend/src/packages/achievements/achievement.repository.ts]:
```suggestion
      .distinctOn('referenceTable')
```

File: backend/src/packages/achievements/achievement.repository.ts [hunk 8]
```diff
@@ -29,14 +40,88 @@ class AchievementRepository implements IRepository {
...
+      .castTo<{ referenceTable: referenceTableType }[]>()
+      .execute();
+
+    const knex = this.achievementModel.knex();
+
+    return await Promise.all(
+      relatedTableNames.map(async ({ referenceTable }) => {
+        const count = (await knex
+          .select()
+          .count(`* as ${referenceTable}`)
```
Comment (andanf-e) [inline backend/src/packages/achievements/achievement.repository.ts]:
```suggestion
          .count('*', { as: referenceTable })
```

File: backend/src/packages/achievements/achievement.repository.ts [hunk 9]
```diff
@@ -29,14 +40,88 @@ class AchievementRepository implements IRepository {
...
+      .select('referenceTable')
+      .groupBy('referenceTable')
+      .castTo<{ referenceTable: referenceTableType }[]>()
+      .execute();
+
+    const knex = this.achievementModel.knex();
+
+    return await Promise.all(
+      relatedTableNames.map(async ({ referenceTable }) => {
+        const count = (await knex
```
Comment (andanf-e) [inline backend/src/packages/achievements/achievement.repository.ts]:
```suggestion
        return knex
```

File: backend/src/packages/achievements/achievement.repository.ts [hunk 10]
```diff
@@ -29,14 +40,88 @@ class AchievementRepository implements IRepository {
...
+    const relatedTableNames = await this.achievementModel
+      .query()
+      .select('referenceTable')
+      .groupBy('referenceTable')
+      .castTo<{ referenceTable: referenceTableType }[]>()
+      .execute();
+
+    const knex = this.achievementModel.knex();
+
+    return await Promise.all(
```
Comment (andanf-e) [inline backend/src/packages/achievements/achievement.repository.ts]:
```suggestion
    const referenceTablesCounts =  await Promise.all(
```

File: backend/src/packages/achievements/achievement.repository.ts [hunk 11]
```diff
@@ -29,14 +40,88 @@ class AchievementRepository implements IRepository {
...
+        const count = (await knex
+          .select()
+          .count(`* as ${referenceTable}`)
+          .from(referenceTable)
+          .where('userId', userId)
+          .first()) as { [key in referenceTableType]: string };
+
+        return { [referenceTable]: Number(count[referenceTable]) };
+      }),
+    );
```
Comment (andanf-e) [inline backend/src/packages/achievements/achievement.repository.ts:123]:
```suggestion
    );
    
    return referenceTablesCounts.reduce<Record<ReferenceTable, number>>((groupedCount, referenceTableToCount) => {
      const [[tableName, count]] = Object.entries(referenceTableToCount)
      return {
        ...groupedCount,
        [tableName]: Number(count)
      };
    }, {});
```

File: backend/src/packages/achievements/achievement.service.ts [hunk 12]
```diff
@@ -31,6 +35,46 @@ class AchievementService implements IService {
...
 
+  public async findOwnProgress(
+    userId: number,
+  ): Promise<AchievementWithProgressResponseDto[]> {
+    const achievements = await this.findAll();
+
+    const userAchievements =
+      await this.achievementRepository.findAllUserAchievements(userId);
+
+    const counts = await this.achievementRepository.getCountsInReferenceTables(
```
Comment (andanf-e) [inline backend/src/packages/achievements/achievement.service.ts]:
```suggestion
    const referenceTableToCount = await this.achievementRepository.getCountsInReferenceTables(
```

File: backend/src/packages/achievements/achievement.service.ts [hunk 13]
```diff
@@ -31,6 +35,46 @@ class AchievementService implements IService {
...
+          progress: PercentageProgress.MAX,
+        };
+      }
+
+      const referenceNameAndCount = counts.find(
+        (count) => Object.keys(count)[0] === achievement.referenceTable,
+      );
+      const count = referenceNameAndCount
+        ? referenceNameAndCount[achievement.referenceTable]
+        : 0;
```
Comment (andanf-e) [inline backend/src/packages/achievements/achievement.service.ts]: After changes with grouping in repository
```suggestion
      const countByAchievementReference = referenceTableToCount[achievement.referenceTable] ?? PercentageProgress.MIN;
```

File: backend/src/packages/achievements/achievement.service.ts [hunk 14]
```diff
@@ -31,6 +35,46 @@ class AchievementService implements IService {
...
+
+      const progress = Math.round((count / achievement.breakpoint) * 100);
```
Comment (andanf-e) [inline backend/src/packages/achievements/achievement.service.ts]:
```suggestion
      const progress = Math.round((count / achievement.breakpoint) * PercentageProgress.MAX);
```

File: frontend/src/pages/profile/components/user-achievements/user-achievements.tsx [hunk 15]
```diff
@@ -32,64 +29,56 @@ type Properties = {
...
-            classNameAchievement={styles.achievementItem}
-          />
-        </div>
-        <Tooltip
-          id={DataTooltipId.ACHIEVEMENT_TOOLTIP}
-          className={styles.achievementTooltip}
+      <Loader isLoading={isLoading} type="dots">
+        <AchievementList
+          achievements={achievementsList.slice(
+            0,
```
Comment (andanf-e) [inline frontend/src/pages/profile/components/user-achievements/user-achievements.tsx]: **B31**

File: readme.md [hunk 16]
```diff
@@ -89,6 +89,8 @@ erDiagram
     varchar key "inner usage key value"
     varchar name "readable name value"
     text description "achivement description"
+    int breakpoint
+    text reference_table
```
Comment (andanf-e) [inline readme.md]:
```suggestion
    text reference_table "referencing to table with records for achievements"
```

**Review by andanf-e — CHANGES_REQUESTED** [review]

**Comment by andanf-e** [comment]
> **I implemented this task in another way.** In this branch I don't use `achievements` from the slice, and `ownAchievements` looks like this: image
> 
> Because I believe, filtering user achievements during calculating progress, and then fetching those user achievements separately from a client side is unnecessary work, and I don't understand where to store `progress` on FE if we would have `state.achievements` - base achievements and `state.ownAchievements` - objects with: id, userId, and achievementId 😐
> 
> * [x]  Create users_achievements (with schema provided from readme) table (user_id | achievement_id cols pair should be unique)
> * [x]   Add 2 columns for achievements table (breakpoint with count of items to achieve it and reference_table with table name for referencing achievement)
> * [x]   On saving Article call achievements service to check wether any achievement breakpoint was accomplished:
> * If breakpoint is accomplished - then make record in users_achievements table
> * [x]   On saving Comment - same way, but for comments table
> * If breakpoint is accomplished - then make record in users_achievements table
> * [x]   On user profile - fetch user achievements and show the actual ones
>   Also implemented calculating achievement progress.
> 
> image
> 
> image

If so - then let's get rid of unused parts from slice

**Comment by hnoievets** [comment]
I would also like to point out that achievements achieved before using the achievement check functionality are displayed incorrectly.

image

**Comment by andanf-e** [comment]
> I would also like to point out that achievements achieved before using the achievement check functionality are displayed incorrectly.
> 
> image

Maybe then let's prepopulate user_achievements table with already achieved for users,
Or on checkAchievement - add greater-than check instead of equality

**Comment by hnoievets** [comment]
> Maybe then let's prepopulate user_achievements table with already achieved for users, Or on checkAchievement - add greater-than check instead of equality

Can I fix it like here https://github.com/BinaryStudioAcademy/bsa-2023-writorium/pull/276/commits/183ddd83af32981ad32060de3c5566e4181846b5?

Or is there a possibility that they will add a new achievement that has already been achieved?

**Comment by andanf-e** [comment]
> > Maybe then let's prepopulate user_achievements table with already achieved for users, Or on checkAchievement - add greater-than check instead of equality
> 
> Can I fix it like here [183ddd8](https://github.com/BinaryStudioAcademy/bsa-2023-writorium/commit/183ddd83af32981ad32060de3c5566e4181846b5)?
> 
> Or is there a possibility that they will add a new achievement that has already been achieved?

This is only display workaround - we do need to have already achieved records in user_achievements table

File: backend/src/db/migrations/20230920200620_record_achieved_achievements_in_users_achievements_table.ts [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+          achievement.referenceTable === tableName &&
+          achievement.breakpoint <= Number(count)
+        ) {
+          await knex(TableName.USERS_ACHIEVEMENTS).insert({
+            userId: user.id,
+            achievementId: achievement.id,
+          });
+        }
+      }
+    }
```
Comment (andanf-e) [inline backend/src/db/migrations/20230920200620_record_achieved_achievements_in_users_achievements_table.ts]: Can we achieve this with raw sql

Comment (hnoievets) [reply backend/src/db/migrations/20230920200620_record_achieved_achievements_in_users_achievements_table.ts]: OK, I'll try

**Review by mhrechkin — DISMISSED** [review]

File: backend/src/db/migrations/20230920200620_record_achieved_achievements_in_users_achievements_table.ts [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+const TableName = {
+  USERS_ACHIEVEMENTS: 'users_achievements',
+  ACHIEVEMENTS: 'achievements',
+  ARTICLES: 'articles',
+  COMMENTS: 'comments',
+} as const;
+
+const up = (knex: Knex): Promise<void> => {
+  return knex.raw(`
+    WITH UsersCounts AS (
```
Comment (andanf-e) [inline backend/src/db/migrations/20230920200620_record_achieved_achievements_in_users_achievements_table.ts]: As CTE is still structure like a table - let's call it as tables
```suggestion
    WITH user_counts AS (
```

File: backend/src/db/migrations/20230920200620_record_achieved_achievements_in_users_achievements_table.ts [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+        SELECT c.user_id, c.id, a.reference_table
+        FROM ${TableName.COMMENTS} c
+        JOIN ${TableName.ACHIEVEMENTS} a ON a.reference_table = '${TableName.COMMENTS}'
+        GROUP BY c.user_id, c.id, a.reference_table
+        UNION ALL
+        SELECT ar.user_id, ar.id, a.reference_table
+        FROM ${TableName.ARTICLES} ar
+        JOIN ${TableName.ACHIEVEMENTS} a ON a.reference_table = '${TableName.ARTICLES}'
+        GROUP BY ar.user_id, ar.id, a.reference_table
+      ) t
```
Comment (andanf-e) [inline backend/src/db/migrations/20230920200620_record_achieved_achievements_in_users_achievements_table.ts]: Let's get rid of shortens, as with them it's hard to understand what table is actually being referenced 
```suggestion
      ) grouped_reference_tables_counts
```

**Review by andanf-e — COMMENTED** [review]
Overall looks nice 👍

File: backend/src/db/migrations/20230920200620_record_achieved_achievements_in_users_achievements_table.ts (hunk 19 above)
Comment (hnoievets) [reply backend/src/db/migrations/20230920200620_record_achieved_achievements_in_users_achievements_table.ts]: Get rid of all the shortens?

**Review by andanf-e — APPROVED** [review]

**Review by mhrechkin — APPROVED** [review]

**Pull request merged**